File: library/src/main/java/com/alamkanak/weekview/model/WeekViewConfig.java
package com.alamkanak.weekview.model;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.util.AttributeSet;
import android.util.TypedValue;

import com.alamkanak.weekview.R;
import com.alamkanak.weekview.drawing.WeekViewDrawingConfig;

import java.util.Calendar;

import static com.alamkanak.weekview.utils.Constants.HOURS_PER_DAY;

public class WeekViewConfig {

    public WeekViewDrawingConfig drawingConfig;

    // Calendar configuration
    public int firstDayOfWeek = Calendar.MONDAY;
    public int numberOfVisibleDays = 3;
    public boolean showFirstDayOfWeekFirst = false;

    // Header bottom line
    public boolean showHeaderRowBottomLine = false;
    public int headerRowBottomLineColor = Color.rgb(102, 102, 102);
    public int headerRowBottomLineWidth = 1;

    // Time column
    public int timeColumnTextColor = Color.BLACK;
    public int timeColumnBackgroundColor = Color.WHITE;
    public int timeColumnPadding = 10;
    public int timeColumnTextSize = 12;

    // Time column separator
    public boolean showTimeColumnSeparator = false;
    public int timeColumnSeparatorColor = Color.rgb(102, 102, 102);
    public int timeColumnSeparatorStrokeWidth = 1;

    // Header row
    public int headerRowTextColor = Color.BLACK;
    public int headerRowBackgroundColor = Color.WHITE;
    public int headerRowPadding = 10;
    public int headerRowTextSize = 12;
    public int todayHeaderTextColor = Color.rgb(39, 137, 228);

    // Event chips
    public int allDayEventHeight = 100;
    public int eventCornerRadius = 0;
    public int eventTextSize = 12;
    public int eventTextColor = Color.BLACK;
    public int eventPadding = 8;

    // Event margins
    public int columnGap = 10;
    public int overlappingEventGap = 0;
    public int eventMarginVertical = 3;
    public int eventMarginHorizontal = 0;

    // Colors
    public int dayBackgroundColor = Color.rgb(255, 255, 255);
    public int todayBackgroundColor = Color.rgb(255, 255, 255);
    public boolean showDistinctWeekendColor = false;
    public boolean showDistinctPastFutureColor = false;
    public int pastBackgroundColor = Color.rgb(227, 227, 227);
    public int futureBackgroundColor = Color.rgb(245, 245, 245);
    public int pastWeekendBackgroundColor = 0;
    public int futureWeekendBackgroundColor = 0;

    // Hour height
    public int hourHeight = 50;
    public int minHourHeight = 0; // no minimum specified (will be dynamic, based on screen)
    public int maxHourHeight = 250;
    public int effectiveMinHourHeight = minHourHeight; // compensates for the fact that you can't keep zooming out.

    // Now line
    public boolean showNowLine = false;
    public int nowLineColor = Color.rgb(102, 102, 102);
    public int nowLineStrokeWidth = 5;

    // Now line dot
    public boolean showNowLineDot = false;
    public int nowLineDotColor = Color.rgb(102, 102, 102);
    public int nowLineDotRadius = 16;

    // Hour separators
    public boolean showHourSeparator = true;
    public int hourSeparatorColor = Color.rgb(230, 230, 230);
    public int hourSeparatorStrokeWidth = 2;

    // Day separators
    public boolean showDaySeparator = true;
    public int daySeparatorColor = Color.rgb(230, 230, 230);
    public int daySeparatorStrokeWidth = 2;

    // Scrolling
    public float xScrollingSpeed = 1f;
    public boolean verticalFlingEnabled = true;
    public boolean horizontalFlingEnabled = true;
    public boolean horizontalScrollingEnabled = true;
    public int scrollDuration = 250;

    public WeekViewConfig(Context context, AttributeSet attrs) {
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.WeekView, 0, 0);
        try {
            // Calendar configuration
            firstDayOfWeek = a.getInteger(R.styleable.WeekView_firstDayOfWeek, firstDayOfWeek);
            numberOfVisibleDays = a.getInteger(R.styleable.WeekView_numberOfVisibleDays, numberOfVisibleDays);
            showFirstDayOfWeekFirst = a.getBoolean(R.styleable.WeekView_showFirstDayOfWeekFirst, showFirstDayOfWeekFirst);

            // Header bottom line
            showHeaderRowBottomLine = a.getBoolean(R.styleable.WeekView_showHeaderRowBottomLine, showHeaderRowBottomLine);
            headerRowBottomLineColor = a.getColor(R.styleable.WeekView_headerRowBottomLineColor, headerRowBottomLineColor);
            headerRowBottomLineWidth = a.getDimensionPixelSize(R.styleable.WeekView_headerRowBottomLineWidth, headerRowBottomLineWidth);

            // Time column
            timeColumnTextColor = a.getColor(R.styleable.WeekView_timeColumnTextColor, timeColumnTextColor);
            timeColumnTextSize = a.getDimensionPixelSize(R.styleable.WeekView_timeColumnTextSize, (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, timeColumnTextSize, context.getResources().getDisplayMetrics()));
            timeColumnPadding = a.getDimensionPixelSize(R.styleable.WeekView_timeColumnPadding, timeColumnPadding);
            timeColumnBackgroundColor = a.getColor(R.styleable.WeekView_timeColumnBackground, timeColumnBackgroundColor);

            // Time column separator
            showTimeColumnSeparator = a.getBoolean(R.styleable.WeekView_showTimeColumnSeparator, showTimeColumnSeparator);
            timeColumnSeparatorColor = a.getColor(R.styleable.WeekView_timeColumnSeparatorColor, timeColumnSeparatorColor);
            timeColumnSeparatorStrokeWidth = a.getDimensionPixelSize(R.styleable.WeekView_timeColumnSeparatorStrokeWidth, timeColumnSeparatorStrokeWidth);

            // Header row
            headerRowTextColor = a.getColor(R.styleable.WeekView_headerRowTextColor, headerRowTextColor);
            headerRowBackgroundColor = a.getColor(R.styleable.WeekView_headerRowBackgroundColor, headerRowBackgroundColor);
            headerRowTextSize = a.getDimensionPixelSize(R.styleable.WeekView_headerRowTextSize, (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, headerRowTextSize, context.getResources().getDisplayMetrics()));
            headerRowPadding = a.getDimensionPixelSize(R.styleable.WeekView_headerRowPadding, headerRowPadding);
            todayHeaderTextColor = a.getColor(R.styleable.WeekView_todayHeaderTextColor, todayHeaderTextColor);

            // Event chips
            allDayEventHeight = a.getDimensionPixelSize(R.styleable.WeekView_allDayEventHeight, allDayEventHeight);
            eventCornerRadius = a.getDimensionPixelSize(R.styleable.WeekView_eventCornerRadius, eventCornerRadius);
            eventTextSize = a.getDimensionPixelSize(R.styleable.WeekView_eventTextSize, (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, eventTextSize, context.getResources().getDisplayMetrics()));
            eventTextColor = a.getColor(R.styleable.WeekView_eventTextColor, eventTextColor);
            eventPadding = a.getDimensionPixelSize(R.styleable.WeekView_eventPadding, eventPadding);

            // Event margins
            columnGap = a.getDimensionPixelSize(R.styleable.WeekView_columnGap, columnGap);
            overlappingEventGap = a.getDimensionPixelSize(R.styleable.WeekView_overlappingEventGap, overlappingEventGap);
            eventMarginVertical = a.getDimensionPixelSize(R.styleable.WeekView_eventMarginVertical, eventMarginVertical);
            eventMarginHorizontal = a.getDimensionPixelSize(R.styleable.WeekView_singleDayHorizontalMargin, eventMarginHorizontal);

            // Colors
            dayBackgroundColor = a.getColor(R.styleable.WeekView_dayBackgroundColor, dayBackgroundColor);
            todayBackgroundColor = a.getColor(R.styleable.WeekView_todayBackgroundColor, todayBackgroundColor);
            showDistinctPastFutureColor = a.getBoolean(R.styleable.WeekView_showDistinctPastFutureColor, showDistinctPastFutureColor);
            showDistinctWeekendColor = a.getBoolean(R.styleable.WeekView_showDistinctWeekendColor, showDistinctWeekendColor);
            pastBackgroundColor = a.getColor(R.styleable.WeekView_pastBackgroundColor, pastBackgroundColor);
            futureBackgroundColor = a.getColor(R.styleable.WeekView_futureBackgroundColor, futureBackgroundColor);
            pastWeekendBackgroundColor = a.getColor(R.styleable.WeekView_pastWeekendBackgroundColor, pastBackgroundColor);
            futureWeekendBackgroundColor = a.getColor(R.styleable.WeekView_futureWeekendBackgroundColor, futureBackgroundColor); // If not set, use the same color as in the week

            // Hour height
            hourHeight = a.getDimensionPixelSize(R.styleable.WeekView_hourHeight, hourHeight);
            minHourHeight = a.getDimensionPixelSize(R.styleable.WeekView_minHourHeight, minHourHeight);
            maxHourHeight = a.getDimensionPixelSize(R.styleable.WeekView_maxHourHeight, maxHourHeight);
            effectiveMinHourHeight = minHourHeight;

            // Now line
            showNowLine = a.getBoolean(R.styleable.WeekView_showNowLine, showNowLine);
            nowLineColor = a.getColor(R.styleable.WeekView_nowLineColor, nowLineColor);
            nowLineStrokeWidth = a.getDimensionPixelSize(R.styleable.WeekView_nowLineStrokeWidth, nowLineStrokeWidth);

            // Now line dot
            showNowLineDot = a.getBoolean(R.styleable.WeekView_showNowLineDot, showNowLineDot);
            nowLineDotColor = a.getColor(R.styleable.WeekView_nowLineDotColor, nowLineDotColor);
            nowLineDotRadius = a.getDimensionPixelSize(R.styleable.WeekView_nowLineDotRadius, nowLineDotRadius);

            // Hour separators
            showHourSeparator = a.getBoolean(R.styleable.WeekView_showHourSeparator, showHourSeparator);
            hourSeparatorColor = a.getColor(R.styleable.WeekView_hourSeparatorColor, hourSeparatorColor);
            hourSeparatorStrokeWidth = a.getDimensionPixelSize(R.styleable.WeekView_hourSeparatorStrokeWidth, hourSeparatorStrokeWidth);

            // Day separators
            showDaySeparator = a.getBoolean(R.styleable.WeekView_showHourSeparator, showDaySeparator);
            daySeparatorColor = a.getColor(R.styleable.WeekView_daySeparatorColor, daySeparatorColor);
            daySeparatorStrokeWidth = a.getDimensionPixelSize(R.styleable.WeekView_daySeparatorStrokeWidth, daySeparatorStrokeWidth);

            // Scrolling
            xScrollingSpeed = a.getFloat(R.styleable.WeekView_xScrollingSpeed, xScrollingSpeed);
            horizontalFlingEnabled = a.getBoolean(R.styleable.WeekView_horizontalFlingEnabled, horizontalFlingEnabled);
            horizontalScrollingEnabled = a.getBoolean(R.styleable.WeekView_horizontalScrollingEnabled, horizontalScrollingEnabled);
            verticalFlingEnabled = a.getBoolean(R.styleable.WeekView_verticalFlingEnabled, verticalFlingEnabled);
            scrollDuration = a.getInt(R.styleable.WeekView_scrollDuration, scrollDuration);
        } finally {
            a.recycle();
        }
    }

    public void setNumberOfVisibleDays(int numberOfVisibleDays) {
        this.numberOfVisibleDays = numberOfVisibleDays;
        drawingConfig.resetOrigin();
    }

    public void setTimeColumnTextSize(int timeColumnTextSize) {
        this.timeColumnTextSize = timeColumnTextSize;
        drawingConfig.setTextSize(timeColumnTextSize);
    }

    public void setTimeColumnTextColor(int textColor) {
        timeColumnTextColor = textColor;
        drawingConfig.setHeaderColumnTextColor(timeColumnTextColor);
    }

    public void setHeaderRowBackgroundColor(int headerRowBackgroundColor) {
        this.headerRowBackgroundColor = headerRowBackgroundColor;
        drawingConfig.headerBackgroundPaint.setColor(headerRowBackgroundColor);
    }

    public void setDayBackgroundColor(int dayBackgroundColor) {
        this.dayBackgroundColor = dayBackgroundColor;
        drawingConfig.dayBackgroundPaint.setColor(dayBackgroundColor);
    }

    public void setTodayBackgroundColor(int todayBackgroundColor) {
        this.todayBackgroundColor = todayBackgroundColor;
        drawingConfig.todayBackgroundPaint.setColor(todayBackgroundColor);
    }

    public void setHourSeparatorStrokeWidth(int hourSeparatorWidth) {
        this.hourSeparatorStrokeWidth = hourSeparatorWidth;
        drawingConfig.hourSeparatorPaint.setStrokeWidth(hourSeparatorWidth);
    }

    public void setTodayHeaderTextColor(int todayHeaderTextColor) {
        this.todayHeaderTextColor = todayHeaderTextColor;
        drawingConfig.todayHeaderTextPaint.setColor(todayHeaderTextColor);
    }

    public void setEventTextSize(int eventTextSize) {
        this.eventTextSize = eventTextSize;
        drawingConfig.eventTextPaint.setTextSize(eventTextSize);
    }

    public void setEventTextColor(int eventTextColor) {
        this.eventTextColor = eventTextColor;
        drawingConfig.eventTextPaint.setColor(eventTextColor);
    }

    public void setTimeColumnBackgroundColor(int timeColumnBackgroundColor) {
        this.timeColumnBackgroundColor = timeColumnBackgroundColor;
        drawingConfig.headerColumnBackgroundPaint.setColor(timeColumnBackgroundColor);
    }

    public float getTotalDayWidth() {
        return drawingConfig.widthPerDay + columnGap;
    }

    public float getTotalDayHeight() {
        float dayHeight = hourHeight * HOURS_PER_DAY;
        float headerHeight = drawingConfig.headerHeight;
        float totalHeaderPadding = headerRowPadding * 2;
        float headerBottomMargin = drawingConfig.headerMarginBottom;
        return dayHeight + headerHeight + totalHeaderPadding + headerBottomMargin;
    }

    public boolean isSingleDay() {
        return numberOfVisibleDays == 1;
    }

    public boolean isWeek() {
        return numberOfVisibleDays == 7;
    }

}

File: core/src/main/java/com/alamkanak/weekview/WeekViewConfig.kt
package com.alamkanak.weekview

import android.content.Context
import android.content.res.TypedArray
import android.graphics.Color
import android.graphics.Typeface
import android.util.AttributeSet

internal class WeekViewConfig(
    context: Context,
    attrs: AttributeSet?
) {

    // Calendar configuration
    var firstDayOfWeek: Int? = null
    var numberOfVisibleDays: Int = 0
    var restoreNumberOfVisibleDays: Boolean = true
    var showFirstDayOfWeekFirst: Boolean = false
    var showCurrentTimeFirst: Boolean = false

    // Header bottom line
    var showHeaderRowBottomLine: Boolean = false
    var headerRowBottomLineColor: Int = 0
    var headerRowBottomLineWidth: Int = 0

    // Time column
    var timeColumnTextColor: Int = 0
    var timeColumnBackgroundColor: Int = 0
    var timeColumnPadding: Int = 0
    var timeColumnTextSize: Int = 0
    var showMidnightHour: Boolean = false
    var showTimeColumnHourSeparator: Boolean = false
    var timeColumnHoursInterval: Int = 0

    // Time column separator
    var showTimeColumnSeparator: Boolean = false
    var timeColumnSeparatorColor: Int = 0
    var timeColumnSeparatorStrokeWidth: Int = 0

    // Header row
    var headerRowTextColor: Int = 0
    var headerRowBackgroundColor: Int = 0
    var headerRowTextSize: Int = 0
    var headerRowPadding: Int = 0
    var todayHeaderTextColor: Int = 0
    var singleLineHeader: Boolean = false

    // Event chips
    var eventCornerRadius: Int = 0
    var eventTextSize: Int = 0
    var adaptiveEventTextSize: Boolean = false
    var eventTextColor: Int = 0

    var eventPaddingHorizontal: Int = 0
    var eventPaddingVertical: Int = 0

    var defaultEventColor: Int = 0
    var allDayEventTextSize: Int = 0

    // Event margins
    var columnGap: Int = 0
    var overlappingEventGap: Int = 0
    var eventMarginVertical: Int = 0
    var eventMarginHorizontal: Int = 0

    // Colors
    var dayBackgroundColor: Int = 0
    var todayBackgroundColor: Int = 0
    var showDistinctWeekendColor: Boolean = false
    var showDistinctPastFutureColor: Boolean = false
    var pastBackgroundColor: Int = 0
    var futureBackgroundColor: Int = 0
    var pastWeekendBackgroundColor: Int = 0
    var futureWeekendBackgroundColor: Int = 0

    // Hour height
    var hourHeight: Float = 0f
    var minHourHeight: Int = 0
    var maxHourHeight: Int = 0
    var effectiveMinHourHeight: Int = 0
    var showCompleteDay: Boolean = false

    // Now line
    var showNowLine: Boolean = false
    var nowLineColor: Int = 0
    var nowLineStrokeWidth: Int = 0

    // Now line dot
    var showNowLineDot: Boolean = false
    var nowLineDotColor: Int = 0
    var nowLineDotRadius: Int = 0

    // Hour separators
    var showHourSeparator: Boolean = false
    var hourSeparatorColor: Int = 0
    var hourSeparatorStrokeWidth: Int = 0

    // Day separators
    var showDaySeparator: Boolean = false
    var daySeparatorColor: Int = 0
    var daySeparatorStrokeWidth: Int = 0

    // Scrolling
    var xScrollingSpeed: Float = 0f
    var verticalFlingEnabled: Boolean = false
    var horizontalFlingEnabled: Boolean = false
    var horizontalScrollingEnabled: Boolean = false
    var scrollDuration: Int = 0

    // Time range
    var minHour: Int = 0
    var maxHour: Int = 0

    // Font
    var typeface: Typeface = Typeface.DEFAULT

    init {
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.WeekView, 0, 0)
        a.use {
            // Calendar configuration
            if (a.hasValue(R.styleable.WeekView_firstDayOfWeek)) {
                firstDayOfWeek = getInteger(R.styleable.WeekView_firstDayOfWeek, 0)
            }

            numberOfVisibleDays = getInteger(R.styleable.WeekView_numberOfVisibleDays, 3)
            restoreNumberOfVisibleDays = getBoolean(R.styleable.WeekView_restoreNumberOfVisibleDays, true)
            showFirstDayOfWeekFirst = getBoolean(R.styleable.WeekView_showFirstDayOfWeekFirst, false)
            showCurrentTimeFirst = getBoolean(R.styleable.WeekView_showCurrentTimeFirst, false)

            // Header bottom line
            showHeaderRowBottomLine = getBoolean(R.styleable.WeekView_showHeaderRowBottomLine, false)
            headerRowBottomLineColor = getColor(R.styleable.WeekView_headerRowBottomLineColor, Defaults.GRID_COLOR)
            headerRowBottomLineWidth = getDimensionPixelSize(R.styleable.WeekView_headerRowBottomLineWidth, 1)

            // Time column
            timeColumnTextColor = getColor(R.styleable.WeekView_timeColumnTextColor, Color.BLACK)
            timeColumnBackgroundColor = getColor(R.styleable.WeekView_timeColumnBackgroundColor, Color.WHITE)
            timeColumnPadding = getDimensionPixelSize(R.styleable.WeekView_timeColumnPadding, 10)
            timeColumnTextSize = getDimensionPixelSize(R.styleable.WeekView_timeColumnTextSize, Defaults.textSize(context))
            showMidnightHour = getBoolean(R.styleable.WeekView_showMidnightHour, false)
            showTimeColumnHourSeparator = getBoolean(R.styleable.WeekView_showTimeColumnHourSeparator, false)
            timeColumnHoursInterval = getInteger(R.styleable.WeekView_timeColumnHoursInterval, 1)

            // Time column separator
            showTimeColumnSeparator = getBoolean(R.styleable.WeekView_showTimeColumnSeparator, false)
            timeColumnSeparatorColor = getColor(R.styleable.WeekView_timeColumnSeparatorColor, Defaults.GRID_COLOR)
            timeColumnSeparatorStrokeWidth = getDimensionPixelSize(R.styleable.WeekView_timeColumnSeparatorStrokeWidth, 1)

            // Time range
            minHour = getInt(R.styleable.WeekView_minHour, 0)
            maxHour = getInt(R.styleable.WeekView_maxHour, 24)

            // Header row
            headerRowTextColor = getColor(R.styleable.WeekView_headerRowTextColor, Color.BLACK)
            headerRowBackgroundColor = getColor(R.styleable.WeekView_headerRowBackgroundColor, Color.WHITE)
            headerRowTextSize = getDimensionPixelSize(R.styleable.WeekView_headerRowTextSize, Defaults.textSize(context))
            headerRowPadding = getDimensionPixelSize(R.styleable.WeekView_headerRowPadding, 10)
            todayHeaderTextColor = getColor(R.styleable.WeekView_todayHeaderTextColor, Defaults.HIGHLIGHT_COLOR)
            singleLineHeader = getBoolean(R.styleable.WeekView_singleLineHeader, true)

            // Event chips
            eventCornerRadius = getDimensionPixelSize(R.styleable.WeekView_eventCornerRadius, 0)
            eventTextSize = getDimensionPixelSize(R.styleable.WeekView_eventTextSize, Defaults.textSize(context))
            adaptiveEventTextSize = getBoolean(R.styleable.WeekView_adaptiveEventTextSize, false)
            eventTextColor = getColor(R.styleable.WeekView_eventTextColor, Color.BLACK)
            defaultEventColor = getColor(R.styleable.WeekView_defaultEventColor, Defaults.EVENT_COLOR)
            allDayEventTextSize = getDimensionPixelSize(R.styleable.WeekView_allDayEventTextSize, eventTextSize)

            // Event padding
            eventPaddingHorizontal = getDimensionPixelSize(R.styleable.WeekView_eventPaddingHorizontal, 8)
            eventPaddingVertical = getDimensionPixelSize(R.styleable.WeekView_eventPaddingVertical, 8)

            // Event margins
            columnGap = getDimensionPixelSize(R.styleable.WeekView_columnGap, 10)
            overlappingEventGap = getDimensionPixelSize(R.styleable.WeekView_overlappingEventGap, 0)
            eventMarginVertical = getDimensionPixelSize(R.styleable.WeekView_eventMarginVertical, 2)
            eventMarginHorizontal = getDimensionPixelSize(R.styleable.WeekView_singleDayHorizontalMargin, 0)

            // Colors
            dayBackgroundColor = getColor(R.styleable.WeekView_dayBackgroundColor, Defaults.BACKGROUND_COLOR)
            todayBackgroundColor = getColor(R.styleable.WeekView_todayBackgroundColor, Defaults.BACKGROUND_COLOR)
            showDistinctPastFutureColor = getBoolean(R.styleable.WeekView_showDistinctPastFutureColor, false)
            showDistinctWeekendColor = getBoolean(R.styleable.WeekView_showDistinctWeekendColor, false)
            pastBackgroundColor = getColor(R.styleable.WeekView_pastBackgroundColor, Defaults.PAST_BACKGROUND_COLOR)
            futureBackgroundColor = getColor(R.styleable.WeekView_futureBackgroundColor, Defaults.FUTURE_BACKGROUND_COLOR)
            pastWeekendBackgroundColor = getColor(R.styleable.WeekView_pastWeekendBackgroundColor, pastBackgroundColor)
            futureWeekendBackgroundColor = getColor(R.styleable.WeekView_futureWeekendBackgroundColor, futureBackgroundColor)

            // Hour height
            hourHeight = getDimension(R.styleable.WeekView_hourHeight, 50f)
            minHourHeight = getDimensionPixelSize(R.styleable.WeekView_minHourHeight, 0)
            maxHourHeight = getDimensionPixelSize(R.styleable.WeekView_maxHourHeight, 400)
            effectiveMinHourHeight = minHourHeight
            showCompleteDay = getBoolean(R.styleable.WeekView_showCompleteDay, false)

            // Now line
            showNowLine = getBoolean(R.styleable.WeekView_showNowLine, false)
            nowLineColor = getColor(R.styleable.WeekView_nowLineColor, Defaults.NOW_COLOR)
            nowLineStrokeWidth = getDimensionPixelSize(R.styleable.WeekView_nowLineStrokeWidth, 5)

            // Now line dot
            showNowLineDot = getBoolean(R.styleable.WeekView_showNowLineDot, false)
            nowLineDotColor = getColor(R.styleable.WeekView_nowLineDotColor, Defaults.NOW_COLOR)
            nowLineDotRadius = getDimensionPixelSize(R.styleable.WeekView_nowLineDotRadius, 16)

            // Hour separators
            showHourSeparator = getBoolean(R.styleable.WeekView_showHourSeparator, true)
            hourSeparatorColor = getColor(R.styleable.WeekView_hourSeparatorColor, Defaults.SEPARATOR_COLOR)
            hourSeparatorStrokeWidth = getDimensionPixelSize(R.styleable.WeekView_hourSeparatorStrokeWidth, 2)

            // Day separators
            showDaySeparator = getBoolean(R.styleable.WeekView_showDaySeparator, true)
            daySeparatorColor = getColor(R.styleable.WeekView_daySeparatorColor, Defaults.SEPARATOR_COLOR)
            daySeparatorStrokeWidth = getDimensionPixelSize(R.styleable.WeekView_daySeparatorStrokeWidth, 2)

            // Scrolling
            xScrollingSpeed = getFloat(R.styleable.WeekView_xScrollingSpeed, 1f)
            horizontalFlingEnabled = getBoolean(R.styleable.WeekView_horizontalFlingEnabled, true)
            horizontalScrollingEnabled = getBoolean(R.styleable.WeekView_horizontalScrollingEnabled, true)
            verticalFlingEnabled = getBoolean(R.styleable.WeekView_verticalFlingEnabled, true)
            scrollDuration = getInt(R.styleable.WeekView_scrollDuration, 250)

            // Font
            val fontFamily = getString(R.styleable.WeekView_fontFamily)
            val typeface = getInteger(R.styleable.WeekView_typeface, 0)
            val textStyle = getInteger(R.styleable.WeekView_textStyle, 0)
            setTypefaceFromAttrs(fontFamily, typeface, textStyle)
        }
    }

    private fun setTypefaceFromAttrs(
        familyName: String?,
        typefaceIndex: Int,
        styleIndex: Int
    ) {
        if (familyName != null) {
            val tf = Typeface.create(familyName, styleIndex)
            if (tf != null) {
                typeface = tf
                return
            }
        }

        typeface = when (typefaceIndex) {
            SANS -> Typeface.SANS_SERIF
            SERIF -> Typeface.SERIF
            MONOSPACE -> Typeface.MONOSPACE
            else -> typeface
        }
    }

    private companion object {
        private const val SANS = 1
        private const val SERIF = 2
        private const val MONOSPACE = 3
    }
}

private fun TypedArray.use(block: TypedArray.() -> Unit) {
    try {
        block()
    } finally {
        recycle()
    }
}
